include/swift/SIL/OptimizationRemark.h

```diff
@@ -0,0 +1,136 @@
+//===--- Emitter.h - Optimization diagnostics -*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the remark type and the emitter class that passes can use
+// to emit optimization diagnostics.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_OPTIMIZATIONREMARKEMITTER_H
+#define SWIFT_SIL_OPTIMIZATIONREMARKEMITTER_H
+
+#include "swift/Basic/SourceLoc.h"
+#include "swift/SIL/SILInstruction.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+
+class ASTContext;
+class SILFunction;
+
+namespace OptRemark {
+
+/// \brief Used in the streaming interface as the general argument type.  It
+/// internally converts everything into a key-value pair.
+struct Argument {
+  std::string Key;
+  std::string Val;
+  // If set, the debug location corresponding to the value.
+  SourceLoc Loc;
+
+  explicit Argument(StringRef Str = "") : Key("String"), Val(Str) {}
+
+  Argument(StringRef Key, int N);
+  Argument(StringRef Key, long N);
+  Argument(StringRef Key, long long N);
+  Argument(StringRef Key, unsigned N);
+  Argument(StringRef Key, unsigned long N);
+  Argument(StringRef Key, unsigned long long N);
+
+  Argument(StringRef Key, SILFunction *F);
+};
+
+/// Shorthand to insert named-value pairs.
+using NV = Argument;
+
+/// The base class for remarks.  This can be created by optimization passed to
+/// report successful and unsuccessful optimizations. CRTP is used to preserve
+/// the underlying type encoding the remark kind in the insertion operator.
+template <typename DerivedT> class Remark {
+  /// Arguments collected via the streaming interface.
+  SmallVector<Argument, 4> Args;
+
+  /// Textual identifier for the remark (single-word, camel-case). Can be used
+  /// by external tools reading the YAML output file for optimization remarks to
+  /// identify the remark.
+  llvm::StringRef Identifier;
+
+  /// Source location for the diagnostics.
+  SourceLoc Location;
+
+protected:
+  Remark(StringRef Identifier, SILInstruction &I)
+      : Identifier(Identifier), Location(I.getLoc().getSourceLoc()) {}
+
+public:
+  DerivedT &operator<<(StringRef S) {
+    Args.emplace_back(S);
+    return *static_cast<DerivedT *>(this);
+  }
+
+  DerivedT &operator<<(Argument A) {
+    Args.push_back(std::move(A));
+    return *static_cast<DerivedT *>(this);
+  }
+
+  SourceLoc getLocation() const { return Location; }
+  std::string getMsg() const;
```

Comment:
Why not return a StringRef?

Reply:
This one is actually function constructing the message from the various arguments that got inserted into the remark.

I can obviously make it a SmallString though.  To answer your other question, I was only using std::string over SmallString because the code in swift I was looking at was also using them exclusively so I though that there was a different policy than in LLVM.

Reply:
Nope. I haven't heard of anything different. @jrose-apple your thoughts?

Reply:
std::string is fine for return values. SmallString is preferred for locals, but I don't think we've ever thought about a perf difference for return values. (It's also a little funny to have the size of the string in a method's API.)

If this is perf sensitive, you could also use a `SmallVectorImpl<char> &` out parameter, which guarantees the "it's on the stack" advantage of SmallString.

Reply:
If the performance of building a string is important, take a raw_ostream.

Reply:
It's unclear at this point.  This interface is also not used when generating remarks so we can change it later if it becomes a performance issue.